---
title: "Jensen-Shannon distance calculations"
author: "Janet Young\n"
date: "`r Sys.Date()`\n"
output: github_document
always_allow_html: true
---

Goal - show how to calculate Jensen-Shannon divergence between two (aligned) alignments at each position    


# Load libraries

```{r setup, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Biostrings)
library(entropy)
```


# Read in example data: short H2A alignment (Antoine)

In this case I have a single alignment of the histone fold domain of various H2A family members. It contains canonical H2A, H2A.B, H2A.L, H2A.P (8 species each) and the marsupial-specific H2A.R (5 sequences) (total of 37 sequences).

I read in the alignment, figure out which H2A variant each sequence is from, and split the single master alignment into 5 individual alignments, one for each H2A variant.

```{r}
aln_file <- "example_alignment_files/exampleProtAln_shortH2As_histoneFoldDomain.fa"

masterAln <- readAAStringSet(aln_file)
# simplify the sequence names by removing the description
names(masterAln) <- sapply(strsplit(names(masterAln), " "), "[[", 1)
# class(masterAln)
# [1] "AAStringSet"
# attr(,"package")
# [1] "Biostrings"

masterAln
```

Then for each sequence, I figure out which variant it's from, and I split the alignment into one for each variant.

```{r}
## figure out which variant each sequence is from
masterAlnSeqTypes <- sapply(strsplit(names(masterAln), "_"), "[[", 1)
masterAlnSeqTypes <- gsub("R[12]$","R", masterAlnSeqTypes)
# table(masterAlnSeqTypes)

## split the alignment into those categories. We get a list object containing all 5 alignments
masterAlnSplit <- split(masterAln, masterAlnSeqTypes)

# names(masterAlnSplit)
# [1] "H2A"   "H2A.B" "H2A.L" "H2A.P" "H2A.R"

## I add a sixth alignment, which is the combined B, L and P alignment
masterAlnSplit[["BandLandP"]] <- c(masterAlnSplit[["H2A.B"]], masterAlnSplit[["H2A.L"]], masterAlnSplit[["H2A.P"]])
```




# Define some useful functions

We define a function called getJSD to get [Jensen-Shannon distance](https://en.wikipedia.org/wiki/Jensen%E2%80%93Shannon_divergence), based on info I got from Mike Doud (Bloom lab), now published [here](https://academic.oup.com/mbe/article/32/11/2944/982113#74376742) (see equation 2).

Jensen-Shannon distance quantifies amino acid preference difference between two frequency profiles.  Note that JS **distance** is the square root of JS **divergence**.


```{r}
# myEnt is a function to get Shannon entropy from a single vector of frequencies.
# It's a simple wrapper function for the entropy.empirical function from the entropy package
# freqs is a vector of numbers that sums to 1 representing amino acid frequencies at a given position in a single alignment. In my case it has length 21 (includes gap character, -)
myEnt <- function (freqs) {  
    # I don't know why, but I had to put in this rounding stage before checking freqs sum to 1, otherwise some positions gave me an error
    mySum <- round(sum(freqs), digits=8)
    if(mySum != 1) {
        # cat("sum ",mySum," freqs ", freqs, "\n\n") # for troubleshooting
        stop("\n\nERROR in myEnt function - frequencies should sum to 1, but they don't\n\n")
    }
    entropy.empirical(freqs, "log2") 
}

# getJSD is a function to get JSD comparing two frequency profiles (freqs1 and freqs2). 
getJSD <- function( freqs1, freqs2 ) {
    firstTerm <- myEnt( (freqs1 + freqs2)/2 )
    secondTerm <- (myEnt(freqs1) + myEnt(freqs2))/2
    myResult <- sqrt( firstTerm - secondTerm )
    return(myResult)
}

# The upper bound of two-class JSD is sqrt(log2(2)) = 1
# The upper bound of three-class JSD is sqrt(log2(3)) = 1.258953
#  (that is correct according the wikipedia page)
# The upper bound of the four-class JSD is sqrt(log2(4)) = 1.414214
# I noticed that Wikipedia definition's of two-sample JSD does NOT divide the first term by 2, but Mike Doud's equation 2 does. I'm going to trust Mike Doud's for now.

## getJSD_multiClass is a function that works on any number of frequency profiles
getJSD_multiClass <- function(freqs_list) {
    firstTerm <- Reduce("+", freqs_list) / length(freqs_list)
    firstTerm <- myEnt(firstTerm)
    secondTerm <- sapply(freqs_list, myEnt)
    secondTerm <- sum(secondTerm) / length(freqs_list)
    myResult <- sqrt( firstTerm - secondTerm )
    return(myResult)
}

#### tests
## getJSD and getJSD_multiClass give same output on the two-class case
# getJSD(masterAlnSplit_freqs[["H2A"]][,7], 
#        masterAlnSplit_freqs[["H2A.B"]][,7]) 
# getJSD_multiClass(list(masterAlnSplit_freqs[["H2A"]][,7], 
#                        masterAlnSplit_freqs[["H2A.B"]][,7]) )
## getJSD_3classes (now removed) and getJSD_multiClass give same output on the three-class case
# getJSD_multiClass(list(masterAlnSplit_freqs[["H2A"]][,7], 
#                        masterAlnSplit_freqs[["H2A.B"]][,7],
#                        masterAlnSplit_freqs[["H2A.P"]][,7]) )
# getJSD_3classes(masterAlnSplit_freqs[["H2A"]][,7], 
#                 masterAlnSplit_freqs[["H2A.B"]][,7],
#                 masterAlnSplit_freqs[["H2A.P"]][,7]) 
```



Get amino acid (plus gap) count and frequency matrices for each alignment

First we define a function called `getAlnCounts()` that takes an alignment and returns a count or frequency matrix

```{r}
# a tiny function that makes sure all seqs in an alignment are the same length as each other
checkAlnLengths <- function(aln) {
    if(length(unique(width(aln))) != 1) {
        stop("\n\nERROR - you supplied a ragged alignment (seqs not all the same length)\n\n")
    } else {
        return(TRUE)
    } 
}

## define the letters we want to count
# AA_STANDARD is defined in the Biostrings package and includes the usual 20 amino acids. I want to add the gap character ("-")
myAAtoTabulate <- c("-", AA_STANDARD)

## define the function
getAlnCounts <- function(aln, letters=myAAtoTabulate, as.prob=FALSE) {
    # check for ragged alns (seqs not all the same length)
    checkAlnLengths(aln)
    
    ## for each sequence, get matrix of 0 and 1 representing the letter at each position. Returns a list of matrices, one for each input seq, with num_rows= num aln positions, and num_columns=21 (- plus each AA)  
    countsEachSeq <- lapply(1:length(aln), function(i) {
        letterFrequencyInSlidingView(aln[[i]], view.width = 1, letters=letters)
    })
    
    # if there were letters in the alignment that are not accounted for in the letters argument, the totals won't be correct.
    expectedTotals <- width(aln)[1]
    totalCountsEachSeq <- sapply(countsEachSeq, sum)
    if ( sum(totalCountsEachSeq != expectedTotals) > 0) {
        ## generate an informative error message:
        problem_seqs <- which(totalCountsEachSeq != expectedTotals)
        problem_seq_letters <- aln[problem_seqs] |> 
            as.character() |> 
            strsplit(split="")
        problem_seq_letters <- lapply(problem_seq_letters, function(x) {
            setdiff(x, myAAtoTabulate) |> 
                unique() |> 
                paste0()
        }) |> 
            paste(collapse=",")
        err_msg <- paste0("\n\nERROR - the total counts didn't add up correctly.\n",
                          "These sequences contain unexpected letters: " , 
                          paste(problem_seqs, collapse=","),
                          "\nAnd those letters are: ",
                          problem_seq_letters,
                          "\n\n")
        stop(err_msg)
    }
    
    # get total counts by position - the Reduce function takes a list object and uses the specified function on all the elements
    countTotals <- Reduce("+", countsEachSeq)
    
    # transpose so columns are positions and rows are each letter type
    countTotals <- t(countTotals)
    
    # perhaps get frequencies not counts
    if(as.prob) {
        freqs <- countTotals / colSums(countTotals)
        return(freqs)
    } else {
        return(countTotals)
    }
}
```


Now use that function on each alignment:

```{r}
masterAlnSplit_counts <- lapply( masterAlnSplit, getAlnCounts, letters=myAAtoTabulate)
masterAlnSplit_freqs <- lapply( masterAlnSplit, getAlnCounts, letters=myAAtoTabulate, as.prob=TRUE)
```

Now test out the `getJSD()` function that gets Jensen-Shannon distance.

Here are frequencies for the 7th position in the H2A alignment:
```{r}
masterAlnSplit_freqs[["H2A"]][,7]
```

and in the H2A.B alignment
```{r}
masterAlnSplit_freqs[["H2A.B"]][,7]
```

here's the corresponding JSD:
```{r}
getJSD (masterAlnSplit_freqs[["H2A"]][,7], masterAlnSplit_freqs[["H2A.B"]][,7] )
```

Same thing for the 2nd position, where H2A and H2A.B are identical - freqs in the H2A alignment:
```{r}
masterAlnSplit_freqs[["H2A"]][,2]
```

and in the H2A.B alignment
```{r}
masterAlnSplit_freqs[["H2A.B"]][,2]
```

here's the corresponding JSD:
```{r}
getJSD (masterAlnSplit_freqs[["H2A"]][,2], masterAlnSplit_freqs[["H2A.B"]][,2] )
```

Now we make a wrapper function that gets JSD at every position between two alignments

```{r}

# I use the getJSD_multiClass in case I want to do it with >2 alignments
getJSDeachPos <- function(aln_list, letters=myAAtoTabulate) {
    # check none of the alignments are ragged
    null <- lapply(aln_list, checkAlnLengths)
    # check all the alns are the same length
    aln_widths <- sapply(aln_list, function(aln) { width(aln)[1] })
    if(length(unique(aln_widths))>1) {
        stop("\n\nThe alignments are not all the same length as each other\n\n")
    }
    # get freq matrices
    freqs_list <- lapply(aln_list, getAlnCounts, letters=letters, as.prob=TRUE)
    # for each column, get the JSDs
    allJSDs <- sapply(1:aln_widths[1], function(i) {
        freqs_this_position <- lapply(freqs_list, function(eachFreqMatrix) {
            eachFreqMatrix[,i]
        })
        jsd <- getJSD_multiClass(freqs_this_position)
        return(jsd)
    })
    return(allJSDs)
}

## test
# temp <- getJSDeachPos(list(masterAlnSplit[["H2A"]], 
#                            masterAlnSplit[["H2A.B"]],
#                            masterAlnSplit[["H2A.P"]]))
```


```{r}
## we'll keep out JSD results in a tibble for later usage
jsd_tbl <- tibble(pos = 1:width(masterAln)[1],
                  aa_in_H2A = strsplit(as.character(masterAln[["H2A_mouse"]]), split="")[[1]]) |> 
    unite("xaxis_label", c(pos, aa_in_H2A), sep="", remove=FALSE)

jsd_tbl$H2AvsB <- getJSDeachPos(list(masterAlnSplit[["H2A"]], masterAlnSplit[["H2A.B"]]))
jsd_tbl$H2AvsLBP <- getJSDeachPos(list(masterAlnSplit[["H2A"]], masterAlnSplit[["BandLandP"]]))
jsd_tbl$H2AvsLvsBvsP <- getJSDeachPos(list(masterAlnSplit[["H2A"]], 
                                           masterAlnSplit[["H2A.B"]], 
                                           masterAlnSplit[["H2A.L"]], 
                                           masterAlnSplit[["H2A.P"]]))
```


Make a plot

```{r, fig.height=2, fig.width=11}
jsd_tbl |> 
    ggplot(aes(x=pos, y=H2AvsLBP)) +
    geom_col(fill="lightgrey", col="darkgray", width=1) +
    theme_classic() +
    labs(x="alignment position, and amino acid in mouse H2A",
         y="JSD, H2A vs H2A.B+L+P") +
    theme(axis.text.x=element_text(angle = 90, hjust=1, vjust=0.5)) + 
    scale_x_continuous(breaks=jsd_tbl$pos, labels=jsd_tbl$xaxis_label)
```



# Finished

show R version used, and package versions

```{r sessionInfo}
sessionInfo()
```


