---
title: "trees_usingPublicTrees"
author: "Janet Young\n"
date: "`r Sys.Date()`\n"
output: github_document
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(tidyverse)
library(Biostrings)
library(snakecase)
library(taxize)
library(here)
library(ggtree)
```


# Goals

I'm trying the [taxize package](https://docs.ropensci.org/taxize/index.html) to interact with taxonomic databases. 

It can help convert latin species names to common names (and vice versa?)

It can use taxonomic classifications to infer trees.

I think it might be able to help [resolve ambiguous names](https://docs.ropensci.org/taxize/articles/name_cleaning.html) 


# get example latin names for some species (mammals)

get example species names I want a tree for. This is long-winded but it should be a working example. 

I read in an alignment, and do a bit of processing to get a bunch of species names (latin).  Show the first 5 rows of an info table I'm building up

```{r}
RTL3_alnFile <- here("Rscripts/phylogenetics/RTL3_aln2022summer_v28.fa")

RTL3_aln <- readDNAStringSet(RTL3_alnFile)
names(RTL3_aln) <- sapply(strsplit(names(RTL3_aln), " "), "[[", 1)

## make tibble of seqnames
RTL3_alnInfoTable <- tibble(id=names(RTL3_aln))

## get species name from seqnames
RTL3_alnInfoTable <- RTL3_alnInfoTable %>% 
    separate(id, into=c(NA,"species"), extra="drop", remove=FALSE) %>% 
    mutate(species=str_replace(species, "CapsidHMMhit", "")) %>% 
    mutate(species=str_replace(species, "ProteaseORF", "")) %>% 
    mutate(species=str_replace(species, "BetweenORFsRegion", "")) %>% 
    mutate(species=str_to_sentence(to_snake_case(species))) %>% 
    mutate(species=str_replace_all(species,"_"," "))

## get rid of some redundant species for this purpose - get rid of any seq whose name contains "annotationSeq"
RTL3_alnInfoTable <- RTL3_alnInfoTable %>% 
    filter(!grepl("annotationSeq",id))

RTL3_alnInfoTable %>% 
    head(5) %>% 
    kable() %>% 
    kable_styling()
```
get just the species names

```{r}
## first get all 81 species names in the table
query_species <- RTL3_alnInfoTable %>% 
    select(species) %>% 
    deframe() %>% 
    unique()

## make a smaller set of species, for quicker testing:
test_query_species <- sample(query_species, 8)

## make a version of that which contains "bad" species names that won't be found in the database
test_query_species_plusBadOnes <- c(test_query_species, 
                                    "fake species name", 
                                    "fake species name2")
```

There were `r length(query_species)` unique species names.

# Look up common names

Now I use the `sci2comm()` function from the `taxize` package to get common names given latin names as input. It's a bit slow, and very noisy.

```{r, include=FALSE}
commonNameLookup <- sci2comm(sci=query_species)
## I checked out the result - it is a list, containing 1 character item found for each query, and none are NA:
# class(commonNameLookup)
# table( sapply(commonNameLookup, length))
# table( sapply(commonNameLookup, is.na))
```

Use that list to add the common names to the tibble I'm building up

```{r}
## add commonNameLookup to RTL3_alnInfoTable
commonNameLookup_tbl <- commonNameLookup %>% 
    unlist() %>% 
    as_tibble(rownames="species") %>% 
    dplyr::rename("common_name"="value")
RTL3_alnInfoTable <- left_join(RTL3_alnInfoTable, commonNameLookup_tbl, by="species") %>% 
    relocate(common_name, .after="species")
RTL3_alnInfoTable %>% 
    head(5) %>% 
    kable() %>% 
    kable_styling()
```

# try taxize classification() function

Look at the `classification()` function of taxize package, using the NCBI taxonomy database:

It returns a list-like object of class "classification". The names are the same as the query species names we provided.  Each element is a data.frame, containing 3 columns (name, rank, taxID) and a bunch of rows (each level of the taxonomic classification). Exception - if the species name was not found, it's `NA` (single element).

```{r try taxize, eval=TRUE}
test_query_species_classification <- classification(test_query_species_plusBadOnes,
                                                    db="ncbi")
# numSpeciesNotFound <- sum(sapply(test_query_species_classification, function(x) {!class(x)=="data.frame"}))
```

# try taxize class2tree to get a tree

Using the results of `classification()`, we can ask get a tree using the `class2tree()` function

```{r}
test_query_species_classification_tree <- class2tree(test_query_species_classification)
plot(test_query_species_classification_tree)

# ## ggtree didn't like this tree. Maybe because it includes a polytomy? Maybe because there's some weird conflicts between packages and functions?
# test_query_species_classification_tree %>%
#     ggtree() +
#     geom_tiplabel()
```

Get the full tree (all species), using NCBI classifications.

```{r, fig.height=11, fig.width=5}
## get full tree - NCBI classifications
specNames_classification_ncbi <- classification(query_species, db="ncbi")
# check how many species not found:
# numSpeciesNotFound <- sum(sapply(specNames_classification, function(x) {!class(x)=="data.frame"}))
# which(sapply(specNames_classification, function(x) {!class(x)=="data.frame"}))
## using ncbi database, all are found. Tree is not always well-resolved
specNames_tree_ncbi <- class2tree(specNames_classification_ncbi)
plot(specNames_tree_ncbi, cex=0.5, main="tree using NCBI taxonomy")
```

Try using encyclopedia of life classifications instead of NCBI. I can get classifications OK (ish) but the class2tree function hangs.

```{r}
# temp2 <- get_eolid("Pinus contorta", ask=FALSE, rows=1) 
# temp1 <- get_eolid("Pinus contorta") 
# get_eolid(query_species[1:2], ask=FALSE, rows=1) 

# specNames_classification_eol <- classification(query_species[1:2], db="eol", rows=1)
specNames_classification_eol <- classification(query_species, db="eol", rows=1)

## I was previous sometimes having trouble with this, getting error:
# Error: Bad Gateway (HTTP 502)
# and no result

## 65 are data.frame and 16 are character. The character ones all look like this "No hierarchy information for 22212420" (with different IDs)

# sapply(specNames_classification_eol, class) %>% table()
# specNames_classification_eol[which(sapply(specNames_classification_eol, class) =="character")]

## get just the 65 data.frame items:
specNames_classification_eol_good <- specNames_classification_eol[ which(sapply(specNames_classification_eol, class)=="data.frame") ]

## still get an error with class2tree because there are some NA. Try getting rid of those:
specNames_classification_eol_good <- lapply(specNames_classification_eol_good, function(x) {
    x %>% 
        filter(!is.na(rank))
})
## some had nothing left after getting rid of the NA.  49 remain after I get rid of those:
specNames_classification_eol_good <- specNames_classification_eol_good[which(sapply(specNames_classification_eol_good, nrow)>0)]

## the class2tree function gets stuck (hangs) at the "Align taxonomy hierarchies..." step on those 49, perhaps because some species don't have enough info (e.g. human, which has only two rows)
# specNames_tree_eol <- class2tree(specNames_classification_eol_good)
# specNames_classification_eol_good[["Homo sapiens"]]
#           name    rank       id
# 1         Homo   genus 13332486
# 2 Homo sapiens species 13332485

## I'll try getting rid of any species that don't have "class". now we have 35 things
specNames_classification_eol_good <- specNames_classification_eol_good[  which(sapply(specNames_classification_eol_good, function(x) { "class" %in% x$rank } )) ]
    
## can get a tree from those 35? no - it still hangs at the "Align taxonomy hierarchies..." step.  I'm not going to try to figure out why.
# specNames_tree_eol <- class2tree(specNames_classification_eol_good)
# plot(specNames_tree_eol, cex=0.5, main="tree using encyclopedia of life taxonomy")
```

Try using Integrated Taxonomic Information Service (itis) classifications


```{r}
## turns out this is slightly tricky - doesn't behave the same way on all queries. the accepted=TRUE parameter seems important (I first thought I needed rows=3 or something, but I think the accepted parameter takes care of that)

# it initially stalls on "Chinchilla lanigera", and I fixed that by changing rows to rows=3.  Turns out there are 2 tsn (like taxID?) for Chinchilla, first one invalid. Using rows>=2 allows the second (valid) ID to be used. 

specNames_classification_itis <- classification(query_species, db="itis", accepted=TRUE)

## Ochotona curzoniae needs rows=1 (or NA)
# classification("Ochotona curzoniae", db="itis", rows=1)
# classification("Ochotona curzoniae", db="itis", rows=2)
# temp_itis <- classification("Ochotona curzoniae", db="itis", rows=10)

## Chinchilla lanigera needs rows=2 (or more)
# classification("Chinchilla lanigera", db="itis", rows=1)
# classification("Chinchilla lanigera", db="itis", rows=2)
# aha! this works
# classification("Chinchilla lanigera", db="itis", accepted=TRUE)
# itis_acceptname("Chinchilla lanigera")

## check how many species not found:
# table(sapply(specNames_classification_itis, class))
# 4 were not found:
# which(sapply(specNames_classification_itis, function(x) {!class(x)=="data.frame"}))
# Pongo pygmaeus abelii       Aotus nancymaae    Nannospalax galili    Cricetulus griseus 
#                     5                    24                    46                    55 
```

I then tried to get fancy and get classifications first from ITIS then from NCBI. Not sure why I bothered.  I'm not running this code, but here's what I tried:

```{r, eval=FALSE}
### use itis when possible, if not, use NCBI
# there's more advice here: https://docs.ropensci.org/taxize/articles/name_cleaning.html
getClassification <- function(oneSpecName) {
    thisClassification <- classification(oneSpecName, db="itis", accepted=TRUE)
    # return(thisClassification)
    if(class(thisClassification[[oneSpecName]]) != "data.frame") {
        cat ("   using NCBI for species ",oneSpecName,"\n")
        thisClassification <- classification(oneSpecName, db="ncbi", accepted=TRUE)
    }
    return(thisClassification)
}
specNames_classification_itisNCBI <- lapply(query_species, getClassification)
names(specNames_classification_itisNCBI) <- query_species
## check how many species not found - all are now found
# table(sapply(specNames_classification_itisNCBI, class))
# table(sapply(specNames_classification_itisNCBI, function(x) {class(x[[1]])}))
# numSpeciesNotFound <- sum(sapply(specNames_classification_itisNCBI, function(x) {!class(x)=="data.frame"}))
# which(sapply(specNames_classification_itisNCBI, function(x) {!class(x)=="data.frame"}))

# myDim <- t(sapply(specNames_classification_itisNCBI, function(x) {dim(x[[1]])}))

specNames_classification_itisNCBI_reformat <- lapply(specNames_classification_itisNCBI, "[[", 1)
attr(specNames_classification_itisNCBI_reformat,"class") <- "classification"
attr(specNames_classification_itisNCBI_reformat,"db") <- "ncbi"

## using ncbi database, all are found. Tree is not always well-resolved
specNames_tree_itisNCBI <- class2tree(specNames_classification_itisNCBI_reformat)
plot(specNames_tree_itisNCBI)
```
