---
title: "miscellaneous_testCode"
author: "Janet Young\n"
date: "`r Sys.Date()`\n"
output: github_document
always_allow_html: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# basic tidyverse
```{r}
m <- mtcars %>% as_tibble()

## show counts in each category
m %>% 
    count(cyl)
```

## barplots of counts in each category
```{r}
m %>% 
    ggplot(aes(x=cyl)) +
    geom_bar()
```

## tribble - a way to manually create small tibbles, row-wise
```{r}
tribble(
    ~colA, ~colB,
    "a",   1,
    "b",   2,
    "c",   3
)
# same as:
tibble(colA=c("A","B","C"), 
       colB=1:3)
```

## the purrr::map functions are a bit like lapply / apply
```{r}
1:10 %>%
    map(rnorm, n = 10)


# do something to each column
mtcars %>% map_dbl(sum)

# split into list, then map.  
# map_dbl simplify output, in this case to a dbl (or numeric)
# same for map_lgl(), map_int() and map_chr()
mtcars %>%
    split(.$cyl) %>%
    map(~ lm(mpg ~ wt, data = .x)) %>%
    map(summary) %>%
    map_dbl("r.squared")

# map_dfr tries to return a data.frame (binding by rows)  (map_dfc is similar but binds by columns)
mtcars %>%
    split(.$cyl) %>%
    map(~ lm(mpg ~ wt, data = .x)) %>%
    map_dfr(~ as.data.frame(t(as.matrix(coef(.)))))
```

# some tidbit for Phoebe's gtf/gff question

```{r}
library(rtracklayer)
?GFFFile
```

# variable names - using variables

## bang-bang

"bang-bang" , a.ka. !!  performs "name injection"

it's explained more [here](https://dplyr.tidyverse.org/articles/programming.html#name-injection)

To see more info: ?rlang::`!!`

here I use it in dplyr::rename - we use `!!` to interpret the variable and then we have to use `:=` (instead of =)
```{r}
newVarName <- "sepalLen_new"
iris %>% 
    dplyr::rename(!!newVarName := Sepal.Length) %>% 
    head()
```

There's also something called ["bang-bang-bang"](https://www.reddit.com/r/Rlanguage/comments/g5m5bh/what_does_the_bang_bang_bang_do/?rdt=45180)

## embracing operator

The 'embracing' operator (`{{ }}`) is useful when we want to pass in variable(s) that we want to be interpreted before being used. It's related to !! and !!!

more discussion [here](https://rlang.r-lib.org/reference/embrace-operator.html),  [here](https://rlang.r-lib.org/reference/topic-data-mask.html) and  [here](https://adv-r.hadley.nz/quasiquotation.html)

The embracing operator is [similar to](https://www.r-bloggers.com/2019/07/bang-bang-how-to-program-with-dplyr/), but not identical to !! ("bang-bang"). It's also [similar to](https://www.reddit.com/r/Rlanguage/comments/g5m5bh/what_does_the_bang_bang_bang_do/?rdt=45180), but not identical to !!! ("bang-bang-bang")

Example:

```{r}
### this wouldn't work
## first define a function
get_var0 <- function(data, column, value) {
    data %>% filter(column == value)
}
## then use it - gives an error this way
# get_var0(mtcars, cyl, 6)
#> Error: Problem with `filter()` input `..1`.
#> x object 'cyl' not found
#> i Input `..1` is `column == value`.

## gives empty tbl output this way (wrong, should be 7):
# get_var0(mtcars, "cyl", 6)

### this DOES work
get_var1 <- function(data, column, value) {
    data %>% filter({{ column }} == value)
}
get_var1(mtcars, cyl, 6)
```

### !!! example: 

from [here](https://www.reddit.com/r/Rlanguage/comments/g5m5bh/what_does_the_bang_bang_bang_do/?rdt=45180)

To see more info: ?rlang::`!!!`

```{r}
## Let's say we want to select the first 3 columns of a data frame. So you can do something like this:
test_df <- tibble(a = 1, b = 1, c = 1, d = 1)
test_df %>%
    select(1, 2, 3)
```

```{r}
## Easy enough. Now let's say we have a list of values that we want to use to replicate the code above. Now if you pass this to the select function it fails:
our_list <- list(1, 2, 3)
# test_df %>%
#     select(our_list)
## That's because that code essentially translates to this, which doesn't work.
# test_df %>%
#     select(list(1, 2, 3))

# What we need to do is "unpack" the list using !!!.
test_df %>%
    select(!!!our_list)
# which translates to this:
# test_df %>%
#     select(1, 2, 3)
```


# adding metadata / attributes

I had a use case where I wanted to associate a few bits of info with a tibble of data. The actual example is that my tibble contained genome-wide t-test results, and I wanted to record eaxctly what method I'd used to do the t-tests)

Note that `attr()` and `atrributes()` are different functions


```{r}
# make example tibble:
a <- 1:5
b <- tibble(a, a * 2)

### add attributes under the methods tag (there can be:
attr(b, "methods") <- "here's some text about the method"
attr(b, "details") <- "here's some more detail about the method"

## access the metadata:
attr(b, "methods")
```

```{r}
## show all attributes:
attributes(b)
```



