---
title: "SRA_query_methods"
author: "Janet Young\n"
date: "`r Sys.Date()`\n"
output: github_document
always_allow_html: true
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# Goal

There are a few R packages/tools that SHOULD help me query SRA and other NCBI databases. Understand what they are.

Caveat - I remember exploring one of these packages in detail once before, and found it worked BUT it used data that was very much out of date. It did not use live queries to NCBI. Which one was it?  I think this was for RNA-seq analysis I was doing for Antoine or Pravrutha.

See ~/public_databases/NCBI/SRA/janet_play_with_tools_to_get_metadata/SRA_methods_NOTES.txt - there's more there.


## Packages

[GEOmetadb](https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html) "an attempt to make access to the metadata associated with samples, platforms, and datasets much more feasible. This is accomplished by parsing all the NCBI GEO metadata into a SQLite database that can be stored and queried locally. GEOmetadb is simply a thin wrapper around the SQLite database along with associated documentation. Finally, the SQLite database is updated regularly as new data is added to GEO and can be downloaded at will for the most up-to-date metadata."

[GEOquery](https://www.bioconductor.org/packages/release/bioc/vignettes/GEOquery/inst/doc/GEOquery.html)

[SRAdb](https://www.bioconductor.org/packages/release/bioc/html/SRAdb.html) "SRAdb is an attempt to make access to the metadata associated with submission, study, sample, experiment and run much more feasible. This is accomplished by parsing all the NCBI SRA metadata into a SQLite database that can be stored and queried locally. Fulltext search in the package make querying metadata very flexible and powerful. fastq and sra files can be downloaded for doing alignment locally. Beside ftp protocol, the SRAdb has funcitons supporting fastp protocol (ascp from Aspera Connect) for faster downloading large data files over long distance. The SQLite database is updated regularly as new data is added to SRA and can be downloaded at will for the most up-to-date metadata."      
* At some point I noted that the metadata was NOT being updated but maybe there are now alternative sources? https://github.com/seandavi/SRAdb/issues/38

I tested it in Dec 2021 - it didn't seem to work very well back then. ~/public_databases/NCBI/SRA/janet_play_with_tools_to_get_metadata/SRAdb_JYdemo.R

[SRAdbV2](https://github.com/seandavi/SRAdbV2) - I don't think this is release-ready.  I don't think it's being updated

[biodbNcbi](https://www.bioconductor.org/packages/release/bioc/vignettes/biodbNcbi/inst/doc/biodbNcbi.html) "provides access to the NCBI databases CCDS, Gene, Pubchem Comp and Pubchem Subst, using biodb package framework." It does look like you can use efetch/esearch/einfo but I think maybe just on a limited set of databases?

[rBLAST](https://github.com/mhahsler/rBLAST) "includes interfaces to blastn, blastp, blastx, and makeblastdb. The BLAST software needs to be downloaded and installed separately."
